compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParserErrorHandler.java

```diff
@@ -3733,6 +3746,9 @@ private ParserRuleContext getNextRuleInternal(ParserRuleContext currentCtx, int
             case RIGHT_DOUBLE_ARROW:
                 // Assumption: RIGHT_DOUBLE_ARROW is only occurs in match clauses
                 // in expr-func-body, it is used by a different alias.
+                if (getParentContext() == ParserRuleContext.CHECKING_EXPRESSION) {
+                    return ParserRuleContext.ERROR_CONSTRUCTOR;
```

Comment:
At the end of error constructor we need to end our current context.
 
Since we can have error constructor in two places in the below grammar(bold places), there's ambiguity. So in order to distinguish at which point we are at, seems we need to start context for `on-fail-check` as well. 

checking-keyword **expression** [on-keyword fail-keyword variable-name double-right-arrow **error-constructor**]

Reply:
Need to end those context at the end of error-constructor. 
i.e. here:
https://github.com/ballerina-platform/ballerina-lang/blob/b965b48532f098d2b3d322253b057a109ca0532b/compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParserErrorHandler.java#L3566-L3567